File: tasks/money-rollup/python/solution/fx.py
"""Directed currency rate graph and exact conversion factors."""

import os
from fractions import Fraction

from money import parse_rate


def build_graph(currencies, rates):
    edges = {}
    for rate in rates:
        if not isinstance(rate, dict) or set(rate) != {"from", "to", "rate"}:
            raise ValueError("malformed rate")
        source, target = rate["from"], rate["to"]
        if source not in currencies or target not in currencies:
            raise ValueError("unknown rate currency")
        if source == target:
            raise ValueError("self rate")
        if (source, target) in edges:
            raise ValueError("duplicate rate")
        edges[(source, target)] = parse_rate(rate["rate"])
    return edges


def factor(edges, source, target):
    """Exact product along the unique shortest directed path."""
    if source == target:
        return Fraction(1)
    if os.environ.get("LAB_SABOTAGE") == "direct-rate-only":
        if (source, target) not in edges:
            raise ValueError("no conversion path")
        return edges[(source, target)]
    reached = {source: (Fraction(1), 1)}
    frontier = [source]
    while frontier:
        following = {}
        for node in frontier:
            base, count = reached[node]
            for (start, end), rate in edges.items():
                if start != node or end in reached:
                    continue
                seen = following.get(end)
                if seen is None:
                    following[end] = (base * rate, count)
                else:
                    following[end] = (seen[0], seen[1] + count)
        if not following:
            break
        reached.update(following)
        if target in reached:
            value, count = reached[target]
            if count != 1:
                raise ValueError("ambiguous conversion path")
            return value
        frontier = list(following)
    raise ValueError("no conversion path")

File: tasks/money-rollup/typescript/environment/src/fx.ts
import { parseRate } from "./money";

export type Graph = Map<string, number>;

function edgeKey(source: string, target: string): string {
  return source + " " + target;
}

export function buildGraph(
  currencies: Record<string, number>,
  rates: Array<Record<string, string>>
): Graph {
  const edges: Graph = new Map();
  for (const rate of rates) {
    edges.set(edgeKey(rate.from, rate.to), parseRate(rate.rate));
  }
  return edges;
}

export function factor(edges: Graph, source: string, target: string): number {
  if (source === target) {
    return 1;
  }
  const direct = edges.get(edgeKey(source, target));
  if (direct === undefined) {
    throw new Error("no conversion rate");
  }
  return direct;
}
